### PR TITLE
Add resource for integrations

### PR DESCRIPTION
**Pull request by asafalima**

### CODE REVIEW

File: internal/provider/resource_integration.go [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+		return
+	}
+
+	model, diagnostics := r.convertToModel(ctx, integration)
+	if len(diagnostics) > 0 {
+		resp.Diagnostics.Append(diagnostics...)
+		return
+	}
+
+	tflog.Trace(ctx, "updated integration", map[string]interface{}{
```
Comment (ofir76) [inline internal/provider/resource_integration.go:310]: Debug

File: internal/provider/resource_integration.go [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+		paramName := param.GetId()
+		paramPossibleValues := param.GetValues()
+		paramDefaultValue := param.GetDefault()
+
+		attributePath := path.Root(fmt.Sprintf(`metadata["%s"]`, paramName))
+
+		metadataValue, hasValue := metadataElements[paramName]
+		if !hasValue {
+			if paramDefaultValue != "" {
+				resp.Diagnostics.AddAttributeError(
```
Comment (ofir76) [inline internal/provider/resource_integration.go:418]: Why do we send an error if we have a default value?

Comment (ofir76) [reply internal/provider/resource_integration.go:418]: can't we use the default value?

File: internal/provider/resource_integration.go [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+
+	// Save imported data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &model)...)
+
+	tflog.Debug(ctx, "imported integration", map[string]interface{}{
+		"id": integrationId,
+	})
+}
+
+func (r *integrationResource) ValidateConfig(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
```
Comment (ofir76) [inline internal/provider/resource_integration.go:378]: so cool!

File: internal/provider/resource_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	var integrations = map[string]*apono.Integration{}
+	httpmock.RegisterResponder(http.MethodPost, "http://api.apono.dev/api/v2/integrations", func(req *http.Request) (*http.Response, error) {
+		var createReq apono.CreateIntegration
+		if err := json.NewDecoder(req.Body).Decode(&createReq); err != nil {
+			return httpmock.NewStringResponse(400, err.Error()), nil
+		}
+
+		id, err := uuid.NewUUID()
+		if err != nil {
+			return httpmock.NewStringResponse(500, err.Error()), nil
```
Comment (ofir76) [inline internal/provider/resource_integration_test.go:28]: don't we want to return the error in this step?

File: internal/provider/resource_integration_test.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	httpmock.RegisterResponder(http.MethodGet, `=~^http://api\.apono\.dev/api/v2/integrations/([^/]+)\z`, func(req *http.Request) (*http.Response, error) {
+		id := httpmock.MustGetSubmatch(req, 1) // 1=first regexp submatch
+		integration, exists := integrations[id]
+		if !exists {
+			return httpmock.NewStringResponse(404, "Integration not found"), nil
+		}
+
+		resp, err := httpmock.NewJsonResponse(200, integration)
+		if err != nil {
+			return httpmock.NewStringResponse(500, err.Error()), nil
```
Comment (ofir76) [inline internal/provider/resource_integration_test.go:59]: sane

File: internal/provider/resource_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		}
+
+		integration.Name = updateReq.Name
+		integration.ProvisionerId = updateReq.ProvisionerId
+		integration.Metadata = updateReq.Metadata
+		integration.SecretConfig = updateReq.SecretConfig
+
+		resp, err := httpmock.NewJsonResponse(200, integration)
+		if err != nil {
+			return httpmock.NewStringResponse(500, err.Error()), nil
```
Comment (ofir76) [inline internal/provider/resource_integration_test.go:85]: same

**Review by ofir76 — APPROVED** [review]
Amazing!!!!
please see the comments :)

**Review by ofir76 — APPROVED** [review]
Amazing!!!!
please see the comments :)

File: internal/provider/provider.go [hunk 7]
```diff
@@ -30,21 +40,83 @@ func (p *AponoProvider) Metadata(_ context.Context, _ provider.MetadataRequest,
...
+				"the APONO_PERSONAL_TOKEN environment variable or provider "+
+				"configuration block personal_token attribute.",
+		)
+	}
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	cfg := apono.NewConfiguration()
```
[the remark this answers is not in the document]
Comment (asafalima) [reply internal/provider/provider.go:102]: We set the host and schema in the following lines, this overrides the servers list of the configuration

File: internal/provider/resource_integration.go [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+		metadata[name] = attrValueToString(value)
+	}
+
+	var secretConfig map[string]interface{}
+	if !data.AwsSecret.IsNull() {
+		attributes := data.AwsSecret.Attributes()
+		secretConfig = map[string]interface{}{
+			"type":      "AWS",
+			"region":    attrValueToString(attributes["region"]),
+			"secret_id": attrValueToString(attributes["secret_arn"]),
```
[the remark this answers is not in the document]
Comment (asafalima) [reply internal/provider/resource_integration.go:267]: Changed to secret_id to match api

File: internal/provider/resource_integration.go (hunk 2 above)
Comment (asafalima) [reply internal/provider/resource_integration.go:418]: No, terraform will complain about diff between the user params and the actual params

File: internal/provider/resource_integration_test.go (hunk 4 above)
Comment (asafalima) [reply internal/provider/resource_integration_test.go:28]: No, we want to return the error so it wound be printed in the provider error

File: internal/provider/resource_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+			ProvisionerId: createReq.ProvisionerId,
+			Status:        apono.INTEGRATIONSTATUS_ACTIVE,
+			Metadata:      createReq.Metadata,
+			SecretConfig:  createReq.SecretConfig,
+		}
+		integrations[integration.Id] = &integration
+
+		resp, err := httpmock.NewJsonResponse(200, integration)
+		if err != nil {
+			return httpmock.NewStringResponse(500, err.Error()), nil
```
[the remark this answers is not in the document]
Comment (asafalima) [reply internal/provider/resource_integration_test.go:44]: no

**Pull request merged**